### PR TITLE
Group Calls (WIP)

### PR DESCRIPTION
**Pull request by robertlong**
# Group Calls

This PR expands upon Matrix calls and adds support for establishing calls to multiple members in a room. The code stems from my work on [matrix-video-chat](https://github.com/vector-im/matrix-video-chat).

### `GroupCall`

Created by calling `client.createGroupCall(roomId, type)` where type can be either `"voice"` or `"video"`. This class essentially sets up a bunch of `MatrixCall`s to room members who have set a participant state event.

### `GroupCallParticipant`

This is a wrapper around `MatrixCall` that represents the state for a participant who has entered a `GroupCall`.

### Addition of the `invitee` field to `m.call.invite`

This `invitee` field allows the recipients of the `m.call.invite` event to filter out invites not intended for them.

### Ice Disconnected Timeout

Rather than wait on the Ice `failed` connection state, I added a 30 second timeout which ensures that calls hang up if you've been disconnected for too long. If the connection state changes within this 30 second period, this timeout resets.

### Active Speaker

@SimonBrandner recently added speaking and volume events to `CallFeed`s, this PR uses that work to track the "active speaker" in a group call. This is primarily intended to be used to show what user should be focused in a group call.

### WebRTC Datachannels

Adding datachannels to calls is currently possible by calling `call.peerConn.createDataChannel()`. This PR just formalizes that process by adding a public API method for doing so and a `datachannel` event for listening for new datachannels on a `MatrixCall` or `GroupCall`.

## To Do:
- [x] Split out unrelated code into separate PRs
- [x] Refactor `getLocalAudioStream`/ `getLocalVideoStream`
- [x] Revert `CallFeed` export
- [ ] Figure out final conference call join/leave/presence events (device events etc.)
- [ ] Write an MSC for the new group call events and `invitee` field.
- [x] Fix known regression with microphone/video mute state
- [ ] Add docs for all new public methods/strings/classes
- [ ] Write release notes

---
This PR currently has no changelog labels, so will not be included in changelogs.

A reviewer can add one of: `T-Deprecation`, `T-Enhancement`, `T-Defect`, `T-Task` to indicate what type of change this is, or add `Type: [enhancement/defect/task]` to the description and I'll add them for you.

### CODE REVIEW

File: src/client.ts [hunk 1]
```diff
@@ -1258,15 +1261,68 @@ export class MatrixClient extends EventEmitter {
         this.supportsCallTransfer = support;
     }
 
+    public async getLocalVideoStream() {
```
Comment (robertlong) [inline src/client.ts:1264]: This is an artifact of the [original group call implementation](https://github.com/vector-im/matrix-video-chat/blob/4d7e5583eb055d0fa2940ac0201f344e70cab588/src/Room.jsx#L232) in matrix-video-chat. I used this method to preview the user's local video stream. I think with the addition of `groupCall.initLocalParticipant()` these public methods are no longer necessary. However, we should still keep track of the `localAVStream`. We want to reuse the `localAVStream` and call `localAVStream.clone()` to get the streams to use in additional calls. In Safari, without calling `localAVStream.clone()` you'll get a black video.

File: src/client.ts [hunk 2]
```diff
@@ -6107,11 +6163,11 @@ export class MatrixClient extends EventEmitter {
     public register(
         username: string,
         password: string,
-        sessionId: string,
+        sessionId: string | null,
```
Comment (robertlong) [inline src/client.ts:6166]: Note for myself, these type changes should be split out into another PR. I need them for matrix-video-call.

File: src/matrix.ts [hunk 3]
```diff
@@ -51,8 +51,14 @@ export {
     createNewMatrixCall,
     setAudioInput as setMatrixCallAudioInput,
     setVideoInput as setMatrixCallVideoInput,
+    CallType,
 } from "./webrtc/call";
 
+// TODO: This export is temporary and is only used for the local call feed for conference calls
+// Ideally conference calls will become a first-class concept and we will have a local call feed with
+// a lifecycle that matches the conference call, not individual calls to members.
+export { CallFeed } from "./webrtc/callFeed";
```
Comment (robertlong) [inline src/matrix.ts:60]: Note for myself. Now that GroupCall has it's own `localParticipant` `CallFeed` we don't need to export this.

File: src/webrtc/call.ts [hunk 4]
```diff
@@ -697,7 +712,14 @@ export class MatrixCall extends EventEmitter {
             this.waitForLocalAVStream = true;
 
             try {
-                const mediaStream = await navigator.mediaDevices.getUserMedia(constraints);
+                let mediaStream: MediaStream;
+
+                if (this.type === CallType.Voice) {
+                    mediaStream = await this.client.getLocalAudioStream();
```
Comment (robertlong) [inline src/webrtc/call.ts:718]: If we remove the public methods for getting local audio stream, we need to add some private method/state in its place.

Comment (robertlong) [reply src/webrtc/call.ts:718]: Maybe `localAVStream` / `localAVStreamPromise` could just be moved to a variable within this module? We only ever want to reference it from the call class. Also then we don't need to export `getUserMediaContraints`

File: src/webrtc/callFeed.ts [hunk 5]
```diff
@@ -188,4 +189,8 @@ export class CallFeed extends EventEmitter {
             this.volumeLooper();
         }, POLLING_INTERVAL);
     }
+
+    public dispose(): void {
+        clearTimeout(this.volumeLooperTimeout);
```
Comment (robertlong) [inline src/webrtc/callFeed.ts:194]: This dispose method just makes sure we don't keep the timeout alive after we've removed the call feed. I can split this out into another PR if we'd like.

File: src/webrtc/groupCall.ts [hunk 6]
```diff
@@ -0,0 +1,762 @@
...
+export enum GroupCallEvent {
+    Entered = "entered",
+    Left = "left",
+    ActiveSpeakerChanged = "active_speaker_changed",
+    ParticipantsChanged = "participants_changed",
+    LocalMuteStateChanged = "local_mute_state_changed",
+}
+
+const CONF_ROOM = "me.robertlong.conf";
+const CONF_PARTICIPANT = "me.robertlong.conf.participant";
```
Comment (robertlong) [inline src/webrtc/groupCall.ts:22]: TODO: I need to move this to the proper MSC event once we've created it.

File: src/webrtc/groupCall.ts [hunk 7]
```diff
@@ -0,0 +1,762 @@
...
+            ACTIVE_SPEAKER_INTERVAL,
+        );
+    };
+
+    /**
+     * Utils
+     */
+
+    // TODO: move this elsewhere or get rid of the retry logic. Do we need it?
+    sendStateEventWithRetry(
```
Comment (robertlong) [inline src/webrtc/groupCall.ts:732]: I don't actually know that we still want this for this PR. I don't think it was actually making calls any more stable, but I don't know for sure.

File: src/webrtc/call.ts [hunk 8]
```diff
@@ -366,9 +370,15 @@ export class MatrixCall extends EventEmitter {
     public async placeVideoCall(): Promise<void> {
         logger.debug("placeVideoCall");
         this.checkForErrorListener();
-        const constraints = getUserMediaContraints(ConstraintsType.Video);
         this.type = CallType.Video;
-        await this.placeCallWithConstraints(constraints);
+        await this.placeCallWithConstraints(ConstraintsType.Video);
+    }
+
+    public createDataChannel(label: string, options: RTCDataChannelInit) {
```
Comment (robertlong) [inline src/webrtc/call.ts:377]: This method and corresponding event could be split into another PR if desired. I would like to reserve the last two parameters to `createGroupCall` for enabling datachannels for all calls and datachannel options though.

File: src/webrtc/call.ts (hunk 4 above)
Comment (robertlong) [reply src/webrtc/call.ts:718]: Spent some time trying to figure out a better way to do this and I couldn't come up with a great solution. We actually want to call `getLocalAudioStream` from both `call.ts` and `groupCall.ts` so it can't be local to `call.ts`. I guess I need to know if it's alright to leave these as public methods or not.

File: src/client.ts (hunk 2 above)
Comment (robertlong) [reply src/client.ts]: Split out into https://github.com/matrix-org/matrix-js-sdk/pull/1903

File: src/webrtc/callFeed.ts (hunk 5 above)
Comment (robertlong) [reply src/webrtc/callFeed.ts:194]: Moved to https://github.com/matrix-org/matrix-js-sdk/pull/1904

File: src/client.ts (hunk 1 above)
Comment (SimonBrandner) [reply src/client.ts:1264]: I think it would make sense to move this to the [`MediaHandler`](https://github.com/SimonBrandner/matrix-js-sdk/commit/a1f73eee86aff8b5b3c667573b44942c3c4e23f9) which will be introduced by https://github.com/matrix-org/matrix-js-sdk/pull/1911

File: src/webrtc/call.ts (hunk 8 above)
Comment (SimonBrandner) [reply src/webrtc/call.ts:377]: I would argue for splitting this out as this PR is already large enough and the data channel stuff feels like a separate thing

File: src/webrtc/call.ts (hunk 4 above)
Comment (SimonBrandner) [reply src/webrtc/call.ts:718]: As mentioned above the [`MediaHandler`](https://github.com/SimonBrandner/matrix-js-sdk/commit/a1f73eee86aff8b5b3c667573b44942c3c4e23f9) from https://github.com/matrix-org/matrix-js-sdk/pull/1911 feels like the way to go to me

File: src/webrtc/groupCall.ts [hunk 9]
```diff
@@ -0,0 +1,565 @@
...
+                sessionId,
+                call,
+            );
+            // TODO: Should we wait until the call has been answered to push the participant?
+            // Or do we hide the participant until their stream is live?
+            // Does hiding a participant without a stream present a privacy problem because
+            // a participant without a stream can still listen in on other user's streams?
+            this.participants.push(participant);
+            this.reEmitter.reEmit(participant, Object.values(GroupCallParticipantEvent));
+            this.emit(GroupCallEvent.ParticipantsChanged), this.participants;
```
Comment (SimonBrandner) [inline src/webrtc/groupCall.ts:495]: This looks weird

**Review by SimonBrandner — COMMENTED** [review]
Generally, this looking absolutely awesome! I'll do a more thorough review when this isn't a draft but I've got a few more general points:

+ [ ] It might be beneficial for `localParticipant` to be a getter the same way `localUserMediaStream` and similar are
+ [x] `GroupCall` should emit `feeds_changed` events and have a `getFeeds()` method which is more friendly for making `CallView` components. It might even allow us to make the already existing `CallView` in the `react-sdk` work without many changes
+ [ ] The TypeScript could be improved a bit (params and return types)
+ [x] [`MediaHandler`](https://github.com/SimonBrandner/matrix-js-sdk/commit/a1f73eee86aff8b5b3c667573b44942c3c4e23f9) from https://github.com/matrix-org/matrix-js-sdk/pull/1911 should be somehow used when it gets merged

File: src/webrtc/call.ts (hunk 8 above)
Comment (robertlong) [reply src/webrtc/call.ts:356]: Started here https://github.com/matrix-org/matrix-js-sdk/pull/1929

File: src/webrtc/call.ts (hunk 4 above)
Comment (robertlong) [reply src/webrtc/call.ts]: I moved everything to use #1911 and #1928

File: src/webrtc/groupCall.ts (hunk 9 above)
Comment (robertlong) [reply src/webrtc/groupCall.ts:489]: You're talking about the comment right? I could use some feedback on that.

File: src/webrtc/groupCall.ts [hunk 10]
```diff
@@ -0,0 +1,565 @@
...
+    Left = "left",
+    ActiveSpeakerChanged = "active_speaker_changed",
+    ParticipantsChanged = "participants_changed",
+    LocalMuteStateChanged = "local_mute_state_changed",
+}
+
+const CONF_ROOM = "me.robertlong.conf";
+const CONF_PARTICIPANT = "me.robertlong.conf.participant";
+
+export class GroupCall extends EventEmitter {
```
[the remark this answers is not in the document]
Comment (robertlong) [reply src/webrtc/groupCall.ts:25]: Yeah, I'll document all the methods before this goes out. Added to the PR's todo list.

File: src/webrtc/groupCall.ts (hunk 9 above)
Comment (SimonBrandner) [reply src/webrtc/groupCall.ts:489]: I meant it probably should be like this, right? 

```suggestion
            this.emit(GroupCallEvent.ParticipantsChanged, this.participants);
```

Or is this some syntax, I am not aware of :sweat_smile:

File: src/webrtc/groupCall.ts [hunk 11]
```diff
@@ -0,0 +1,559 @@
...
+            participant = new GroupCallParticipant(
+                this,
+                member,
+                sessionId,
+                call,
+            );
+            // TODO: Should we wait until the call has been answered to push the participant?
+            // Or do we hide the participant until their stream is live?
+            // Does hiding a participant without a stream present a privacy problem because
+            // a participant without a stream can still listen in on other user's streams?
```
Comment (SimonBrandner) [inline src/webrtc/groupCall.ts:486]: If we end up doing this using `m.call.join` events in the way, I talked about, I think, we can push to participants when we start the call, without waiting for the answer, though I don't think we should show any feeds for them at that point. 

I've been actually thinking that it could be interesting trying to push new feeds once we receive `SDPStreamMetadata` and give `CallFeed` a loading state but I guess that is a topic for another time

Comment (SimonBrandner) [reply src/webrtc/groupCall.ts:486]: We could even try to send `m.call.sdp_stream_metadata_changed` before the actual negotiation to get the info about new loading streams to the other side more quickly though I am not sure how much real benefit would that have. 

I am going to try to do some experiments with this

File: src/webrtc/groupCall.ts (hunk 9 above)
Comment (robertlong) [reply src/webrtc/groupCall.ts:489]: Ah, good catch haha :sweat_smile:

File: src/webrtc/groupCall.ts [hunk 12]
```diff
@@ -0,0 +1,559 @@
...
+
+    private speakerMap: Map<RoomMember, number[]> = new Map();
+    private presenceLoopTimeout?: number;
+    private activeSpeakerLoopTimeout: number;
+    private reEmitter: ReEmitter;
+
+    constructor(
+        private client: MatrixClient,
+        roomId: string,
+        public type: CallType,
```
Comment (SimonBrandner) [inline src/webrtc/groupCall.ts:44]: We are slowly killing of `CallType`s in `MatrixCall` and I think we should do the same here. This feels even more important for conferences - users should be allowed to join in whatever way they want (no camera, no mic, no nothing, whatever :D)

Comment (SimonBrandner) [reply src/webrtc/groupCall.ts:50]: We've decided to keep this as the initial state in the MSC but I think the js-doc should reflect this is just the initial type

File: src/webrtc/callEventHandler.ts [hunk 13]
```diff
@@ -273,7 +273,6 @@ export class CallEventHandler {
                         " and canceling outgoing call " + existingCall.callId,
                     );
                     existingCall.replacedBy(call);
-                    call.answer();
```
Comment (SimonBrandner) [inline src/webrtc/callEventHandler.ts:215]: Won't this break regular VoIP calls?

File: src/webrtc/call.ts [hunk 14]
```diff
@@ -1997,7 +1997,10 @@ export class MatrixCall extends EventEmitter {
         if (this.opponentDeviceId) {
             return this.client.sendToDevice(eventType, {
                 [this.invitee || this.getOpponentMember().userId]: {
-                    [this.opponentDeviceId]: realContent,
+                    [this.opponentDeviceId]: {
+                        ...realContent,
+                        device_id: this.client.deviceId,
```
Comment (SimonBrandner) [inline src/webrtc/call.ts:2002]: I am wondering if the js-sdk wouldn't allow you to get the deviceId from the `MatrixEvent`, so that you don't have to send it

**Comment by no-1ne** [comment]
Looking forward

Will there be limitations on number of members in the call before quality degrades from the perspective of someone sharing a screen to 10-20 participants...

can getting the participants to act as relay to 1-2 other participants help.

**Comment by SimonBrandner** [comment]
> Will there be limitations on number of participants in the call before quality degrades especially the scenario of someone sharing a screen to 10-20..50...100 participants...

At first, this will work as a full-mesh - all participants call all other participants, which will limit the number of users to ~8, afaik, before the quality starts to get worse, of course, this will depend on your computer and connection. Then the work on an SFU is planned which should solve this and possibly allow hundreds of users in the future if done well. See https://github.com/matrix-org/matrix-doc/pull/3401

**Comment by no-1ne** [comment]
SFU makes sense scaling wise, thought matrix want to avoid anything centralised

**Comment by t3chguy** [comment]
Each call/room/server could define its own SFU to use

**Comment by SimonBrandner** [comment]
> Each call/room/server could define its own SFU to use

Also, there can be multiple SFUs involved in a call

**Comment by no-1ne** [comment]
Aiming for 64-128 participants, Can participants  act as a relay...each relay  node forming a 2-4 full mesh. There by presenter to 4 nodes, those 4 nodes to 4 other nodes. Next hop reaching 64 participants. 

The reason being, Not many has access to a publicly accessible droplet to create a Mediasoup instance

**Comment by SimonBrandner** [comment]
> Aiming for 64-128 participants, Can participants act as a relay...each relay node forming a 2-4 full mesh. There by presenter to 4 nodes, those 4 nodes to 4 other nodes. Next hop reaching 64 participants.
> 
> The reason being, Not many has access to a publicly accessible droplet to create a Mediasoup instance

This isn't implemented atm but might be in the future

File: src/webrtc/call.ts [hunk 15]
```diff
@@ -630,6 +630,11 @@ export class MatrixCall extends EventEmitter {
      * @param {boolean} addToPeerConnection whether to add the tracks to the peer connection
      */
     public pushLocalFeed(callFeed: CallFeed, addToPeerConnection = true): void {
+        if (this.feeds.some((feed) => callFeed.stream.id === feed.stream.id)) {
```
Comment (SimonBrandner) [inline src/webrtc/call.ts:633]: Does the same stream id always mean the same feed?

File: src/webrtc/call.ts (hunk 14 above)
Comment (bwindels) [reply src/webrtc/call.ts:2004]: Indeed, you should be able to find out through the sender key of the olm decryption (unsure if/how the js-sdk exposes that though), which has the added benefit that it can't be spoofed. See https://github.com/matrix-org/matrix-doc/pull/3401#discussion_r805893936

**Comment by dbkr** [comment]
Closing in favour of https://github.com/matrix-org/matrix-js-sdk/pull/2355
